cumulus/xcm/xcm-emulator/src/lib.rs

```diff
@@ -351,6 +350,8 @@ macro_rules! decl_test_relay_chains {
 				pallets = {
 					$($pallet_name:ident: $pallet_path:path,)*
 				}
+				on_initialize = $($on_initialize_fn:expr,)?
+				$(on_finalize = $on_finalize_fn:expr,)?
```

Comment:
Makes sense, this was the missing piece I think, thanks for your help here.

Reply:
I've updated my PR with the changes, it's cleaner with the trait but I can't arbitrarily call the hooks within the macro. I need to create an appropriate context for these hooks to be called. I'd imagine `impl OnHooks` within each `decl` macro.

Reply:
The idea is to create a struct for each particular case to feed the macro `hooks` attribute.

The struct will impl `OnHooks`.

```rust
impl OnHooks for MyHooks { .. }
```
```rust
	pub struct Westend {
		genesis = westend::genesis(),
		on_init = (),
		runtime = westend_runtime,
		core = {
			SovereignAccountOf: westend_runtime::xcm_config::LocationConverter, //TODO: rename to SovereignAccountOf,
		},
		pallets = {
			XcmPallet: westend_runtime::XcmPallet,
			Sudo: westend_runtime::Sudo,
			Balances: westend_runtime::Balances,
			Treasury: westend_runtime::Treasury,
			AssetRate: westend_runtime::AssetRate,
		},
		hooks = MyHooks // or `()`
	},
```

Because for most cases hooks will not be needed, `OnHooks` should be impl also for `()`